modules/javafx.base/src/test/java/test/com/sun/javafx/event/CompositeEventHandlerTest.java

```diff
@@ -45,6 +48,132 @@ public void setUp() {
         compositeEventHandler = new CompositeEventHandler<Event>();
     }
 
+    /**
+     * test state report after add/clear weak filter
+     * Here we test that a garbage collected weak filter is actually
+     * removed from the chain.
+     */
+    @Test
+    public void testHasFilterWeakCleared() {
+        final EventCountingHandler<Event> eventCountingHandler =
+                new EventCountingHandler<Event>();
+        final WeakEventHandler<Event> weakEventHandler =
+                new WeakEventHandler<Event>(eventCountingHandler);
+
+        compositeEventHandler.addEventFilter(weakEventHandler);
+        assertFalse("must not have handler after adding filter", compositeEventHandler.hasHandler());
+        assertTrue("must have filter", compositeEventHandler.hasFilter());
+        WeakEventHandlerUtil.clear(weakEventHandler);
+        assertFalse("must not have filter", compositeEventHandler.hasFilter());
+        assertFalse("must not have handler", compositeEventHandler.hasHandler());
+    }
+
+    /**
+     * test state report after add/clear weak handler
+     */
+    @Test
+    public void testHasHandlerAddWeakClear() {
+        final EventCountingHandler<Event> eventCountingHandler =
+                new EventCountingHandler<Event>();
+        final WeakEventHandler<Event> weakEventHandler =
+                new WeakEventHandler<Event>(eventCountingHandler);
+        compositeEventHandler.addEventHandler(weakEventHandler);
+        assertTrue("sanity: really added?", CompositeEventHandlerShim.containsHandler(
+                compositeEventHandler, weakEventHandler));
```

Comment:
I see this hear, but not in the filter test. Is there a reason for the difference? I don't have a concern, I'm just curious.

Reply:
good question :)

Just checked and thought I forgot to make filter tests fully symmetric with handler tests (started work on handlers and copied the available asserts over). So tried a quick addition, and stumbled:

- signature of CompositeEventHandlerShim is containsFilter(CompositeEH, WeakEventHandler) - that is not applicable for not-weak. 
- package test api compositeEventHandler.containsHandler only checks for added, not singletonHandler

Options:

- make filter/handler tests consistent by change/add api of shim and add sanity asserts to filter/singleton handler tests
- remove sanity asserts from new handler tests 
- do nothing (my favorite :)

Reply:
> Options:
> ...
>     * do nothing (my favorite :)

I think we have a winner. :)

I'll integrate this some time this morning.